src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt

```diff
@@ -18,16 +21,51 @@ class CombatSimulation(val world: World) {
             .sortedByDescending { (initiativeRoll, _) -> initiativeRoll }
             .map { (_, eachUnit) -> eachUnit }
 
-    fun run() {
+    fun runRound() {
         initiativeOrder.forEach { unit ->
-            unit.tacticalActionStrategy
-                    .decideTurnActions(world, unit, unit.availableActionOptions)
-                    .let { actionsToExecute ->
-                       unit.executeActions(actionsToExecute)
-                    }
-                    .also {
-                        world.removeUnits(world.findDead())
-                    }
+            if((unit.currentAttributes.getValue(Attribute.CURRENT_HEALTH) as NumericValue).value > 0) {
+                // check if unit alive rather than try removing dead units from list as concurrency risks
+                // caller needs to exception handle this in case CURRENT_HEALTH not provided in json
+                unit.tacticalActionStrategy
+                        .decideTurnActions(world, unit, unit.availableActionOptions)
+                        .let { actionsToExecute ->
+                            unit.executeActions(actionsToExecute)
+                        }
+                        .also {
+                            val deadUnits = world.findDead()
```

Comment:
You can avoid having to dedicated `val` statement by starting a function chain.
`world.findDead().let { deadUnits -> ... }.forEach { deadUnit -> ... }`